### PR TITLE
common/Formatter: use CachedStackStringStream for efficiency

### PR DESCRIPTION
**Pull request by batrick**
A stringstream is incredibly expensive to construct, use a cached one.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by batrick** [comment]
jenkins test api

**Comment by batrick** [comment]
jenkins test make check

**Comment by batrick** [comment]
jenkins test make check arm64

File: src/common/Formatter.cc [hunk 1]
```diff
@@ -29,27 +30,39 @@ namespace ceph {
...
-	t.width(scale + 1);
-	t << num;
-	int len = t.str().size();
-	return t.str().substr(0,len - scale) + "." + t.str().substr(len - scale);
+	css->fill('0');
+	css->width(scale + 1);
+	*css << num;
+	auto len = css->strv().size();
+
+        CachedStackStringStream css2;
```
Comment (rzarzynski) [inline src/common/Formatter.cc:40]: nit: mixed indentation?

File: src/common/Formatter.cc [hunk 2]
```diff
@@ -29,27 +30,39 @@ namespace ceph {
...
 
-	t.fill('0');
-	t.width(scale + 1);
-	t << num;
-	int len = t.str().size();
-	return t.str().substr(0,len - scale) + "." + t.str().substr(len - scale);
+	css->fill('0');
+	css->width(scale + 1);
+	*css << num;
+	auto len = css->strv().size();
```
Comment (rzarzynski) [inline src/common/Formatter.cc:38]: I guess we call `strv()` three times. I'm not sure compilers optimize this.

File: src/common/Formatter.cc [hunk 3]
```diff
@@ -29,27 +30,39 @@ namespace ceph {
...
-	t.width(scale + 1);
-	t << num;
-	int len = t.str().size();
-	return (neg ? "-" : "") + t.str().substr(0,len - scale) + "." + t.str().substr(len - scale);
+	css->fill('0');
+	css->width(scale + 1);
+	*css << num;
+	auto len = css->strv().size();
+
+        CachedStackStringStream css2;
```
Comment (rzarzynski) [inline src/common/Formatter.cc:60]: nit: ditto.

**Review by rzarzynski — COMMENTED** [review]
Generally LGTM. Just some minors.

File: src/common/Formatter.cc (hunk 2 above)
Comment (batrick) [reply src/common/Formatter.cc]: strv() is very cheap (`O(1)`) in this case so it doesn't really matter.

File: src/common/Formatter.cc (hunk 3 above)
Comment (batrick) [reply src/common/Formatter.cc]: fixed

File: src/common/Formatter.cc (hunk 1 above)
Comment (batrick) [reply src/common/Formatter.cc]: fixed

**Comment by batrick** [comment]
jenkins test make check

**Comment by batrick** [comment]
jenkins test make check arm64

**Comment by batrick** [comment]
jenkins test api

**Comment by batrick** [comment]
jenkins test api

**Review by rzarzynski — APPROVED** [review]

File: src/common/Formatter.cc (hunk 2 above)
Comment (rzarzynski) [reply src/common/Formatter.cc]: :+1:

**Comment by batrick** [comment]
This PR is under test in [https://tracker.ceph.com/issues/66086](https://tracker.ceph.com/issues/66086).

**Comment by sebastian-philipp** [comment]
Morning @batrick . I think https://tracker.ceph.com/issues/66215 is loosely related?

**Comment by sajibreadd** [comment]
Is there any backport tracker created for this PR?

**Comment by sebastian-philipp** [comment]
> Is there any backport tracker created for this PR?

not sure, it's not a bug fix?

**Comment by sajibreadd** [comment]
> not sure, it's not a bug fix?

My PR https://github.com/ceph/ceph/pull/57695 uses `CachedStackStream` as it is only used in the main branch. reef is still using `stringstream`. As we need this change in reef, so should we cherry-pick both commit or handle this separately after cherry-pick only my commit?

**Pull request merged**